lib/trino-hive-formats/src/main/java/io/trino/hive/formats/ion/IonDecoderFactory.java

```diff
@@ -206,6 +212,61 @@ private void decode(IonReader ionReader, IntFunction<BlockBuilder> blockSelector
         }
     }
 
+    private static class MapDecoder
+            implements BlockDecoder
+    {
+        private final BlockDecoder valueDecoder;
+        private final Type keyType;
+        private final Type valueType;
+        private final DistinctMapKeys distinctMapKeys;
+        private BlockBuilder keyBlockBuilder;
+        private BlockBuilder valueBlockBuilder;
+
+        public MapDecoder(MapType mapType, BlockDecoder valueDecoder)
+        {
+            this.keyType = mapType.getKeyType();
+            this.valueType = mapType.getValueType();
+            this.valueDecoder = valueDecoder;
+            this.distinctMapKeys = new DistinctMapKeys(mapType, true);
+            this.keyBlockBuilder = mapType.getKeyType().createBlockBuilder(null, 128);
+            this.valueBlockBuilder = mapType.getValueType().createBlockBuilder(null, 128);
+        }
+
+        @Override
+        public void decode(IonReader ionReader, BlockBuilder builder)
+        {
+            ionReader.stepIn();
+            // buffer the keys and values
+            while (ionReader.next() != null) {
+                if (keyType instanceof VarcharType _ || keyType instanceof CharType _) {
+                    VarcharType.VARCHAR.writeSlice(keyBlockBuilder, Slices.utf8Slice(ionReader.getFieldName()));
+                }
+                else {
+                    throw new UnsupportedOperationException("Unsupported map key type: " + keyType);
+                }
+                valueDecoder.decode(ionReader, valueBlockBuilder);
+            }
+            ValueBlock keys = keyBlockBuilder.buildValueBlock();
```

Comment:
I'm curious about the approach to build up temp blocks and then append them. Seems like more overhead than just overwriting when one encounters a repeated key. I realize that there is some bookkeeping needed to do the value over writing but I suspect it may actually be simpler to code.

Reply:
A relevant difference vs the json implementations is that they already have a materialized intermediate representation of sorts, whereas we are just working on the document stream. So that may tip the calculus as to what makes sense.

Reply:
I will take look at this again and see if can simplify this.